prov/efa/src/efa_user_info.c

```diff
@@ -332,7 +332,99 @@ int efa_user_info_alter_rdm(int version, struct fi_info *info, const struct fi_i
 }
 
 /**
- * @brief get a list of rdm info the fit user's requirements
+ * @brief update EFA direct info to match user hints
+ *
+ * the input info is a duplicate of prov info, which matches
+ * the capability of the EFA device. This function tailor it
+ * so it matches user provided hints
+ *
+ * @param	version[in]	libfabric API version
+ * @param	info[in,out]	info to be updated
+ * @param	hints[in]	user provided hints
+ * @return	0 on success
+ * 		negative libfabric error code on failure
+ */
+static
+int efa_user_info_alter_direct(int version, struct fi_info *info, const struct fi_info *hints)
+{
+	if (hints && (hints->caps & FI_HMEM)) {
+		/*
+		 * FI_HMEM is a primary capability, therefore only check
+		 * (and cliam) its support when user explicitly requested it.
+		 */
+		if (!efa_user_info_should_support_hmem(version)) {
+			return -FI_ENODATA;
+		}
+
+		info->caps |= FI_HMEM;
+	} else {
+		info->caps &= ~FI_HMEM;
+	}
+
+	if (info->caps & FI_HMEM) {
+		/* Add FI_MR_HMEM to mr_mode when claiming support of FI_HMEM
+		 * because EFA provider's HMEM support rely on
+		 * application to provide descriptor for device buffer.
+		 */
+		if (hints->domain_attr &&
+		    !(hints->domain_attr->mr_mode & FI_MR_HMEM)) {
+			EFA_WARN(FI_LOG_CORE,
+			        "FI_HMEM capability requires device registrations (FI_MR_HMEM)\n");
+			return -FI_ENODATA;
+		}
+
+		info->domain_attr->mr_mode |= FI_MR_HMEM;
+	}
+
+	/*
+	 * Handle user-provided hints and adapt the info object passed back up
+	 * based on EFA-specific constraints.
+	 */
+	if (hints) {
+		/* We only support manual progress for RMA operations */
+		if (hints->caps & FI_RMA) {
+			info->domain_attr->data_progress = FI_PROGRESS_MANUAL;
+		}
+
+		/*
+		 * Same goes for prefix mode, where the protocol does not
+		 * absolutely need a prefix before receive buffers, but it can
+		 * use it when available to optimize transfers with endpoints
+		 * having the following profile:
+		 *	- Requires FI_MSG and not FI_TAGGED/FI_ATOMIC/FI_RMA
+		 *	- Can handle registrations (FI_MR_LOCAL)
+		 *	- No need for FI_DIRECTED_RECV
+		 *	- Guaranteed to send msgs smaller than info->nic->link_attr->mtu
+		 */
+		if (hints->mode & FI_MSG_PREFIX) {
```

Comment:
I am not sure on this. efa-direct rdm doesn't require prefix. If application supports FI_MSG_PREFIX, we can just ignore it and return an zero prefix. Also the current efa_msg.c code doesn't actually handle prefix for rdm ep type.

Reply:
fixed